org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/RecordPattern.java

```diff
@@ -192,11 +192,11 @@ private void setAccessorsPlusInfuseInferredType(BlockScope scope) {
 		} else {
 			for (int i = 0; i < components.length; i++) {
 				Pattern p = this.patterns[i];
-				if (!(p instanceof TypePattern))
+				if (!(p instanceof VariablePattern))
 					continue;
-				TypePattern tp = (TypePattern) p;
```

Comment:
Can you explain what exactly is achieved by introducing a new layer in the class hierarchy here ? I am not sure `VariablePattern` is a suitable name - but can say for sure only after I understand the purpose

Reply:
The main reason I avoided using or subclassing `TypePattern` was that many of the methods expect that the type has a source range, even when it's an elided type. I ended up using many hacks in order to get unused variables working. I'll take another look to see if I can avoid creating a separate class.

Reply:
> The main reason I avoided using or subclassing `TypePattern` was that many of the methods expect that the type has a source range, even when it's an elided type. I ended up using many hacks in order to get unused variables working. I'll take another look to see if I can avoid creating a separate class.

Let us wait for feedback from @jarthana in this. I am not saying what you have done is wrong. Jay let us have your opinion on this. TIA.

Reply:
It's interesting to note that the TypePattern gets its TypeReference from this.local.type, which comes from VariablePattern. So, I am not sure why we can't make it work. If you can tell me what breaks without this change, that'll help.

Reply:
> It's interesting to note that the TypePattern gets its TypeReference from this.local.type, which comes from VariablePattern. So, I am not sure why we can't make it work. If you can tell me what breaks without this change, that'll help. Edit: I see that we deal with a similar case in lambda arguments where the argument's type is set to null when explicit type is provided. We might be able to make do with null?

I played around and was able to make new tests pass without the VariablePattern. Losing the FakeTypeReference, however, might be a bit more involving. So, I am not particular about the latter change.

Other point is, do we need to do the analyseCode() for the unnamed pattern? Currently, UnnamedPattern delegates to the one in TypePattern but not sure we even need to do that.

Reply:
I deleted the UnnamedPattern class entirely and am just using TypePattern. I also deleted VariablePattern and FakeTypeReference.